client/components/infinite-scroll/index.js

```diff
@@ -3,53 +3,92 @@
 /**
  * External dependencies
  */
-import { Component } from 'react';
+import React from 'react';
 import PropTypes from 'prop-types';
-import { throttle } from 'lodash';
+import { throttle, defer } from 'lodash';
+import afterLayoutFlush from 'lib/after-layout-flush';
 
 const SCROLL_CHECK_RATE_IN_MS = 400;
 
-export default class InfiniteScroll extends Component {
+export default class InfiniteScroll extends React.Component {
 	static propTypes = {
 		nextPageMethod: PropTypes.func.isRequired,
 	};
 
-	checkScrollPositionHandler = throttle(
-		() => this.checkScrollPosition( true ),
-		SCROLL_CHECK_RATE_IN_MS
-	);
+	observedElement = React.createRef();
 
 	componentDidMount() {
-		window.addEventListener( 'scroll', this.checkScrollPositionHandler );
-		this.checkScrollPosition( false );
+		this.hasIntersectionObserver = 'IntersectionObserver' in window;
+
+		if ( this.hasIntersectionObserver && this.observedElement.current ) {
+			this.observer = new IntersectionObserver( this.handleIntersection, {
+				rootMargin: '100%',
+				threshold: 1.0,
+			} );
+			this.observer.observe( this.observedElement.current );
+		} else {
+			window.addEventListener( 'scroll', this.checkScrollPositionHandler );
+			this.deferredTimer = defer( () => this.checkScrollPositionHandler( false ) );
+		}
 	}
 
 	componentDidUpdate() {
-		this.checkScrollPosition( false );
+		if ( ! this.hasIntersectionObserver && ! this.deferredTimer ) {
```

Comment:
I think we lost one useful feature of `InfiniteScroll` here: if the page is not fully populated by current items, it fetches next page repeatedly until it gets populated and the page is long enough to scroll.

That's achieved by checking the scroll position in `componentDidUpdate`, which is called when a new page of results is rendered.

With `hasIntersectionObserver`, this check is never performed.

Reply:
Yup, this was missing indeed.

I've reworked the logic around this, so it should now match the behaviour of the scroll event version; `componentDidUpdate` will now keep scheduling new fetches until the anchor element is no longer visible.